### ISSUE TITLE
loadtxt does not accurately save float128 values

### ISSUE THREAD

**Issue by ahaldane** [numpy/numpy#10436]
This is demonstrated by comparing the scalar output from dragon4 to the string output by `loadtxt`.

(output for x64 arch)

```python

[3]: a = np.array([1.0/np.float128(3.0)])
[4]: np.savetxt(sys.stdout, a)
3.333333333333333148e-01
[7]: a[0]
0.33333333333333333334
```

**Comment by eric-wieser** [comment]
`np.savetxt(sys.stdout, ...)` is a neat trick!

**Comment by ahaldane** [comment]
Incidentally, for float64 and lower `loadtxt` outputs more digits than needed:

```
[20]: b = np.array([1.0/np.float64(3.0)])
[21]: np.savetxt(sys.stdout, b)
3.333333333333333148e-01
[22]: b[0]
0.3333333333333333
```

But at least this doesn't lose info since these values are equal for 64 bit floats: 
```python
>>> 3.333333333333333148e-01 == 0.3333333333333333
True
```

**Comment by eric-wieser** [comment]
My guess is that it uses an `%f` format string, which causes a cast to `float`?

**Comment by ahaldane** [comment]
Yeah, seems so.

**Comment by charris** [comment]
See also #7895.

**Comment by ahaldane** [comment]
That's what led me to discover this!

**Comment by eric-wieser** [comment]
Just linked the two issues together - hope people are OK with me editing other peoples posts to add cross references!

**Comment by charris** [comment]
Pushing this one off to 1.15 also.

**Comment by mattip** [comment]
It is not only the formatting string, inside `savetxt` the formatting itself does `fmt % value` which will not print the correct value:
`'%.20lf' % np.float128(1 + np.finfo(np.float128).eps)` => `'1.00000000000000000000'`

`str(np.float128(1 + np.finfo(np.float128).eps))` => `'1.0000000000000000001'`

`loadtxt` does `np.longdouble(txt)` which parses correctly `np.longdouble('1.0000000000000000001') != np.longdouble('1.0')` =>True, so if we could solve the `savetxt` side, the issue could be solved

**Comment by charris** [comment]
@ahaldane Does this need to be pushed off again?

**Comment by charris** [comment]
@ahaldane Did this get fixed with latest merges?

**Comment by ahaldane** [comment]
Yes, if you are planning on releasing 1.15 soon, this should be pushed off. I didn't have time to work on it, and it is a pretty involved fix.

**Comment by charris** [comment]
OK, done. Thanks.

**Comment by charris** [comment]
@ahaldane What is the status of this?

**Comment by ahaldane** [comment]
I haven't worked on it at all.

I did notice recently in #5543 the suggestion to implement `__format__` on numpy arrays (and scalars?). Maybe that improvement could play into this one.

**Comment by charris** [comment]
OK, I'll push this off to 1.17.

**Comment by mattip** [comment]
Removing the milestone, this needs some deep refactoring of savetxt to eliminate the modulo-formatting for `float128`
